### PR TITLE
[refactor](Coordinator) refactor coordinator

### PR DESCRIPTION
**Pull request by 924060929**
## Proposed changes

Use NereidsSqlCoordinator instead of Coordinator because the code of Coordinator is too hard to maintaining

The main design approach is as follows:

1. Divide the original flat Coordinator into multiple modules, with each module maintaining high cohesion.
- `DistributePlanner`: The logic for calculating parallelism has been extracted in #36531, and in the future, we will dynamically calculate parallelism based on cost.
- `CoordinatorContext`: Some global parameters and states related to the Coordinator are encapsulated within CoordinatorContext.
- `PipelineExecutionTask`: The entire scheduling task is encapsulated by PipelineExecutionTask, which includes the mapping relationship between each Backend and Pipeline task. PipelineExecutionTask contains two layers of tasks, each responsible for specific duties, with state maintenance handled internally rather than being centralized in the Coordinator.
  - `MultiFragmentsPipelineTask`: A Backend will generate multiple fragment tasks, which are bundled together and sent concurrently to the corresponding Backend.
  - `SingleFragmentPipelineTask`: A single fragment task for a Backend.
- `JobProcessor`: Describes two types of tasks: SQL tasks and Load tasks.
  - `QueryProcessor`: Represents query tasks and provides a ResultReceiver to obtain query results.
  - `LoadProcessor`: Represents Insert into and Broker load tasks, providing a blocking function to wait for load completion.
- `ThriftPlansBuilder`: Uses the DistributedPlan structure to build thrift parameters and encapsulates some intermediate temporary variables within functions, rather than placing them in the Coordinator.

2. The overall Coordinator logic is more clearly organized. We can see that the NereidsCoordinator consists of only a few functions, allowing quick understanding of the main flow when reading the code.
  - Construct CoordinatorContext.
  - Enqueue the tasks.
  - Handle different sinks accordingly.
  - Register the Coordinator with `QeProcessorImpl` for cancellation and progress tracking.
  - Construct thrift parameters.
  - Build PipelineTask.
  - Initiate RPC calls to each Backend.

TODO:
1. delete old `Coordinator`
2. support cloud mode

### CODE REVIEW

**Comment by 924060929** [comment]
run buildall

**Comment by 924060929** [comment]
run buildall

**Comment by 924060929** [comment]
run buildall

**Comment by 924060929** [comment]
run buildall

**Comment by 924060929** [comment]
run buildall

**Comment by 924060929** [comment]
run buildall

**Comment by 924060929** [comment]
run buildall

**Comment by 924060929** [comment]
run buildall

**Comment by 924060929** [comment]
run buildall

**Comment by 924060929** [comment]
run buildall

**Comment by 924060929** [comment]
run buildall

**Comment by 924060929** [comment]
run buildall

**Comment by 924060929** [comment]
run buildall

**Comment by 924060929** [comment]
run buildall

**Comment by 924060929** [comment]
run buildall

**Comment by 924060929** [comment]
run buildall

**Comment by 924060929** [comment]
run buildall

**Comment by 924060929** [comment]
run buildall

**Comment by 924060929** [comment]
run buildall

**Comment by 924060929** [comment]
run buildall

**Comment by 924060929** [comment]
run buildall

**Comment by 924060929** [comment]
run buildall

**Comment by 924060929** [comment]
run buildall

**Comment by 924060929** [comment]
run buildall

**Comment by 924060929** [comment]
run buildall

**Comment by 924060929** [comment]
run buildall

**Comment by 924060929** [comment]
run buildall

**Comment by morrySnow** [comment]
run buildall

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/distribute/worker/ScanWorkerSelector.java [hunk 1]
```diff
@@ -49,14 +51,14 @@ public interface ScanWorkerSelector {
...
-            UnassignedScanBucketOlapTableJob unassignedJob);
+            UnassignedScanBucketOlapTableJob unassignedJob, ConnectContext context);
 
     static TScanRangeParams buildScanReplicaParams(
             TScanRangeLocations tabletLocation, TScanRangeLocation replicaLocation) {
         TScanRangeParams replicaParam = new TScanRangeParams();
         replicaParam.scan_range = tabletLocation.scan_range;
         // Volume is optional, so we need to set the value and the is-set bit
-        replicaParam.setVolumeId(replicaLocation.volume_id);
+        // replicaParam.setVolumeId(replicaLocation.volume_id);
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/distribute/worker/ScanWorkerSelector.java:61]: remove useless code?

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/commands/insert/InsertIntoTableCommand.java [hunk 2]
```diff
@@ -289,4 +394,55 @@ private boolean childIsEmptyRelation(PhysicalSink sink) {
     public StmtType stmtType() {
         return StmtType.INSERT;
     }
+
+    private static class ExecutorFactory {
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/commands/insert/InsertIntoTableCommand.java:398]: add comment to explain what this class does

**Review by morrySnow — DISMISSED** [review]

**Comment by 924060929** [comment]
run buildall

**Comment by 924060929** [comment]
run buildall

**Comment by 924060929** [comment]
run buildall

**Comment by 924060929** [comment]
run buildall

**Comment by 924060929** [comment]
run buildall

**Comment by 924060929** [comment]
run buildall

**Comment by 924060929** [comment]
run buildall

**Comment by 924060929** [comment]
run buildall

**Review by starocean999 — APPROVED** [review]

**Review by morrySnow — APPROVED** [review]

**Pull request merged**